props/websubmit/src/main.rs

```diff
@@ -85,16 +93,98 @@ impl DeletionProp {
+/// necessary for e.g. the deletion to work.
+pub struct ScopedStorageProp {
+    cx: Arc<PolicyContext>,
+}
+
+impl ScopedStorageProp {
+    pub fn new(cx: Arc<PolicyContext>) -> Self {
+        ScopedStorageProp { cx }
+    }
+
+    pub fn check(self) {
+        for c_id in self.cx.desc().controllers.keys() {
+            let scopes = self
+                .cx
+                .all_nodes_for_ctrl(*c_id)
+                .filter(|node| self.cx.has_marker(marker!(scopes), *node))
+                .collect::<Vec<_>>();
+            let stores = self
+                .cx
+                .all_nodes_for_ctrl(*c_id)
+                .filter(|node| self.cx.has_marker(marker!(stores), *node))
+                .collect::<Vec<_>>();
+            let mut sensitives = self
+                .cx
+                .all_nodes_for_ctrl(*c_id)
+                .filter(|node| self.cx.has_marker(marker!(sensitive), *node));
+
+            let controller_valid = sensitives.all(|sens| {
+                stores.iter().all(|&store| {
+                    // sensitive flows to store implies some scope flows to store callsite
+                    !(self
+                        .cx
+                        .flows_to(sens, store, paralegal_policy::EdgeType::Data))
+                        || store.typ.as_call_site().is_some_and(|cs| {
+                            scopes.iter().any(|scope| {
+                                self.cx.flows_to(
+                                    *scope,
+                                    Node {
+                                        ctrl_id: store.ctrl_id,
+                                        typ: cs.into(),
+                                    },
+                                    paralegal_policy::EdgeType::Data,
+                                )
+                            })
+                        })
+                })
+            });
+
+            assert!(controller_valid);
+        }
+    }
+}
+
 fn main() -> Result<()> {
     let ws_dir = std::env::args()
         .nth(1)
-        .ok_or_else(|| anyhow!("expected an argument"))?;
-    paralegal_policy::SPDGGenCommand::global()
-        .run(ws_dir)?
-        .with_context(|ctx| {
+        .ok_or_else(|| anyhow!("expected format: cargo run <path> [policy]"))?;
+    let prop_name = match std::env::args().nth(2) {
+        Some(n) => n,
+        None => "".to_owned(),
+    };
+
+    let prop: fn(Arc<Context>) -> Result<_> = match prop_name.as_str() {
+        "sc" => |ctx| {
+            ctx.named_policy(Identifier::new_intern("Scoped Storage Policy"), |ctx| {
+                ScopedStorageProp::new(ctx).check();
+                Ok(())
+            })
+        },
+        "del" => |ctx| {
             ctx.named_policy(Identifier::new_intern("Deletion Policy"), |ctx| {
                 DeletionProp::new(ctx).check();
                 Ok(())
             })
-        })
+        },
+        _ => |ctx| {
+            ctx.clone()
```

Comment:
Since this is duplicated I would make separate functions for `sc` and `del` and then in the combined version call them

Reply:
Done